File: src/main/java/com/merge/api/resources/accounting/types/TaxRate.java
package com.merge.api.resources.accounting.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = TaxRate.Builder.class)
public final class TaxRate {
    private final Optional<String> description;

    private final Optional<Double> totalTaxRate;

    private final Optional<Double> effectiveTaxRate;

    private final Optional<TaxRateCompany> company;

    private final Optional<Boolean> remoteWasDeleted;

    private final Optional<String> id;

    private final Optional<String> remoteId;

    private final Optional<OffsetDateTime> modifiedAt;

    private final Optional<Map<String, JsonNode>> fieldMappings;

    private final Optional<List<RemoteData>> remoteData;

    private TaxRate(
            Optional<String> description,
            Optional<Double> totalTaxRate,
            Optional<Double> effectiveTaxRate,
            Optional<TaxRateCompany> company,
            Optional<Boolean> remoteWasDeleted,
            Optional<String> id,
            Optional<String> remoteId,
            Optional<OffsetDateTime> modifiedAt,
            Optional<Map<String, JsonNode>> fieldMappings,
            Optional<List<RemoteData>> remoteData) {
        this.description = description;
        this.totalTaxRate = totalTaxRate;
        this.effectiveTaxRate = effectiveTaxRate;
        this.company = company;
        this.remoteWasDeleted = remoteWasDeleted;
        this.id = id;
        this.remoteId = remoteId;
        this.modifiedAt = modifiedAt;
        this.fieldMappings = fieldMappings;
        this.remoteData = remoteData;
    }

    /**
     * @return The tax rate's description.
     */
    @JsonProperty("description")
    public Optional<String> getDescription() {
        return description;
    }

    /**
     * @return The tax rate's total tax rate.
     */
    @JsonProperty("total_tax_rate")
    public Optional<Double> getTotalTaxRate() {
        return totalTaxRate;
    }

    /**
     * @return The tax rate's effective tax rate.
     */
    @JsonProperty("effective_tax_rate")
    public Optional<Double> getEffectiveTaxRate() {
        return effectiveTaxRate;
    }

    /**
     * @return The company the tax rate belongs to.
     */
    @JsonProperty("company")
    public Optional<TaxRateCompany> getCompany() {
        return company;
    }

    /**
     * @return Indicates whether or not this object has been deleted by third party webhooks.
     */
    @JsonProperty("remote_was_deleted")
    public Optional<Boolean> getRemoteWasDeleted() {
        return remoteWasDeleted;
    }

    @JsonProperty("id")
    public Optional<String> getId() {
        return id;
    }

    /**
     * @return The third-party API ID of the matching object.
     */
    @JsonProperty("remote_id")
    public Optional<String> getRemoteId() {
        return remoteId;
    }

    /**
     * @return This is the datetime that this object was last updated by Merge
     */
    @JsonProperty("modified_at")
    public Optional<OffsetDateTime> getModifiedAt() {
        return modifiedAt;
    }

    @JsonProperty("field_mappings")
    public Optional<Map<String, JsonNode>> getFieldMappings() {
        return fieldMappings;
    }

    @JsonProperty("remote_data")
    public Optional<List<RemoteData>> getRemoteData() {
        return remoteData;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof TaxRate && equalTo((TaxRate) other);
    }

    private boolean equalTo(TaxRate other) {
        return description.equals(other.description)
                && totalTaxRate.equals(other.totalTaxRate)
                && effectiveTaxRate.equals(other.effectiveTaxRate)
                && company.equals(other.company)
                && remoteWasDeleted.equals(other.remoteWasDeleted)
                && id.equals(other.id)
                && remoteId.equals(other.remoteId)
                && modifiedAt.equals(other.modifiedAt)
                && fieldMappings.equals(other.fieldMappings)
                && remoteData.equals(other.remoteData);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                this.description,
                this.totalTaxRate,
                this.effectiveTaxRate,
                this.company,
                this.remoteWasDeleted,
                this.id,
                this.remoteId,
                this.modifiedAt,
                this.fieldMappings,
                this.remoteData);
    }

    @Override
    public String toString() {
        return "TaxRate{" + "description: " + description + ", totalTaxRate: " + totalTaxRate + ", effectiveTaxRate: "
                + effectiveTaxRate + ", company: " + company + ", remoteWasDeleted: " + remoteWasDeleted + ", id: " + id
                + ", remoteId: " + remoteId + ", modifiedAt: " + modifiedAt + ", fieldMappings: " + fieldMappings
                + ", remoteData: " + remoteData + "}";
    }

    public static Builder builder() {
        return new Builder();
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder {
        private Optional<String> description = Optional.empty();

        private Optional<Double> totalTaxRate = Optional.empty();

        private Optional<Double> effectiveTaxRate = Optional.empty();

        private Optional<TaxRateCompany> company = Optional.empty();

        private Optional<Boolean> remoteWasDeleted = Optional.empty();

        private Optional<String> id = Optional.empty();

        private Optional<String> remoteId = Optional.empty();

        private Optional<OffsetDateTime> modifiedAt = Optional.empty();

        private Optional<Map<String, JsonNode>> fieldMappings = Optional.empty();

        private Optional<List<RemoteData>> remoteData = Optional.empty();

        private Builder() {}

        public Builder from(TaxRate other) {
            description(other.getDescription());
            totalTaxRate(other.getTotalTaxRate());
            effectiveTaxRate(other.getEffectiveTaxRate());
            company(other.getCompany());
            remoteWasDeleted(other.getRemoteWasDeleted());
            id(other.getId());
            remoteId(other.getRemoteId());
            modifiedAt(other.getModifiedAt());
            fieldMappings(other.getFieldMappings());
            remoteData(other.getRemoteData());
            return this;
        }

        @JsonSetter(value = "description", nulls = Nulls.SKIP)
        public Builder description(Optional<String> description) {
            this.description = description;
            return this;
        }

        public Builder description(String description) {
            this.description = Optional.of(description);
            return this;
        }

        @JsonSetter(value = "total_tax_rate", nulls = Nulls.SKIP)
        public Builder totalTaxRate(Optional<Double> totalTaxRate) {
            this.totalTaxRate = totalTaxRate;
            return this;
        }

        public Builder totalTaxRate(Double totalTaxRate) {
            this.totalTaxRate = Optional.of(totalTaxRate);
            return this;
        }

        @JsonSetter(value = "effective_tax_rate", nulls = Nulls.SKIP)
        public Builder effectiveTaxRate(Optional<Double> effectiveTaxRate) {
            this.effectiveTaxRate = effectiveTaxRate;
            return this;
        }

        public Builder effectiveTaxRate(Double effectiveTaxRate) {
            this.effectiveTaxRate = Optional.of(effectiveTaxRate);
            return this;
        }

        @JsonSetter(value = "company", nulls = Nulls.SKIP)
        public Builder company(Optional<TaxRateCompany> company) {
            this.company = company;
            return this;
        }

        public Builder company(TaxRateCompany company) {
            this.company = Optional.of(company);
            return this;
        }

        @JsonSetter(value = "remote_was_deleted", nulls = Nulls.SKIP)
        public Builder remoteWasDeleted(Optional<Boolean> remoteWasDeleted) {
            this.remoteWasDeleted = remoteWasDeleted;
            return this;
        }

        public Builder remoteWasDeleted(Boolean remoteWasDeleted) {
            this.remoteWasDeleted = Optional.of(remoteWasDeleted);
            return this;
        }

        @JsonSetter(value = "id", nulls = Nulls.SKIP)
        public Builder id(Optional<String> id) {
            this.id = id;
            return this;
        }

        public Builder id(String id) {
            this.id = Optional.of(id);
            return this;
        }

        @JsonSetter(value = "remote_id", nulls = Nulls.SKIP)
        public Builder remoteId(Optional<String> remoteId) {
            this.remoteId = remoteId;
            return this;
        }

        public Builder remoteId(String remoteId) {
            this.remoteId = Optional.of(remoteId);
            return this;
        }

        @JsonSetter(value = "modified_at", nulls = Nulls.SKIP)
        public Builder modifiedAt(Optional<OffsetDateTime> modifiedAt) {
            this.modifiedAt = modifiedAt;
            return this;
        }

        public Builder modifiedAt(OffsetDateTime modifiedAt) {
            this.modifiedAt = Optional.of(modifiedAt);
            return this;
        }

        @JsonSetter(value = "field_mappings", nulls = Nulls.SKIP)
        public Builder fieldMappings(Optional<Map<String, JsonNode>> fieldMappings) {
            this.fieldMappings = fieldMappings;
            return this;
        }

        public Builder fieldMappings(Map<String, JsonNode> fieldMappings) {
            this.fieldMappings = Optional.of(fieldMappings);
            return this;
        }

        @JsonSetter(value = "remote_data", nulls = Nulls.SKIP)
        public Builder remoteData(Optional<List<RemoteData>> remoteData) {
            this.remoteData = remoteData;
            return this;
        }

        public Builder remoteData(List<RemoteData> remoteData) {
            this.remoteData = Optional.of(remoteData);
            return this;
        }

        public TaxRate build() {
            return new TaxRate(
                    description,
                    totalTaxRate,
                    effectiveTaxRate,
                    company,
                    remoteWasDeleted,
                    id,
                    remoteId,
                    modifiedAt,
                    fieldMappings,
                    remoteData);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/accounting/TaxRate.kt
package dev.merge.api.models.accounting

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import dev.merge.api.models.*
import java.time.OffsetDateTime
import java.util.Objects
import java.util.Optional

/**
 * # The TaxRate Object
 *
 * ### Description
 *
 * The `TaxRate` object is used to represent a tax rate.
 *
 * ### Usage Example
 *
 * Fetch from the `LIST TaxRates` endpoint and view tax rates relevant to a company.
 */
@JsonDeserialize(builder = TaxRate.Builder::class)
@NoAutoDetect
class TaxRate
private constructor(
    private val description: JsonField<String>,
    private val totalTaxRate: JsonField<Double>,
    private val effectiveTaxRate: JsonField<Double>,
    private val company: JsonField<String>,
    private val remoteWasDeleted: JsonField<Boolean>,
    private val id: JsonField<String>,
    private val remoteId: JsonField<String>,
    private val fieldMappings: JsonValue,
    private val modifiedAt: JsonField<OffsetDateTime>,
    private val remoteData: JsonField<List<RemoteData>>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    /** The tax rate's description. */
    fun description(): Optional<String> =
        Optional.ofNullable(description.getNullable("description"))

    /** The tax rate's total tax rate. */
    fun totalTaxRate(): Optional<Double> =
        Optional.ofNullable(totalTaxRate.getNullable("total_tax_rate"))

    /** The tax rate's effective tax rate. */
    fun effectiveTaxRate(): Optional<Double> =
        Optional.ofNullable(effectiveTaxRate.getNullable("effective_tax_rate"))

    /** The company the tax rate belongs to. */
    fun company(): Optional<String> = Optional.ofNullable(company.getNullable("company"))

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    fun remoteWasDeleted(): Optional<Boolean> =
        Optional.ofNullable(remoteWasDeleted.getNullable("remote_was_deleted"))

    fun id(): Optional<String> = Optional.ofNullable(id.getNullable("id"))

    /** The third-party API ID of the matching object. */
    fun remoteId(): Optional<String> = Optional.ofNullable(remoteId.getNullable("remote_id"))

    /** This is the datetime that this object was last updated by Merge */
    fun modifiedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(modifiedAt.getNullable("modified_at"))

    fun remoteData(): Optional<List<RemoteData>> =
        Optional.ofNullable(remoteData.getNullable("remote_data"))

    /** The tax rate's description. */
    @JsonProperty("description") @ExcludeMissing fun _description() = description

    /** The tax rate's total tax rate. */
    @JsonProperty("total_tax_rate") @ExcludeMissing fun _totalTaxRate() = totalTaxRate

    /** The tax rate's effective tax rate. */
    @JsonProperty("effective_tax_rate") @ExcludeMissing fun _effectiveTaxRate() = effectiveTaxRate

    /** The company the tax rate belongs to. */
    @JsonProperty("company") @ExcludeMissing fun _company() = company

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    @JsonProperty("remote_was_deleted") @ExcludeMissing fun _remoteWasDeleted() = remoteWasDeleted

    @JsonProperty("id") @ExcludeMissing fun _id() = id

    /** The third-party API ID of the matching object. */
    @JsonProperty("remote_id") @ExcludeMissing fun _remoteId() = remoteId

    @JsonProperty("field_mappings") @ExcludeMissing fun _fieldMappings() = fieldMappings

    /** This is the datetime that this object was last updated by Merge */
    @JsonProperty("modified_at") @ExcludeMissing fun _modifiedAt() = modifiedAt

    @JsonProperty("remote_data") @ExcludeMissing fun _remoteData() = remoteData

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate(): TaxRate = apply {
        if (!validated) {
            description()
            totalTaxRate()
            effectiveTaxRate()
            company()
            remoteWasDeleted()
            id()
            remoteId()
            modifiedAt()
            remoteData().map { it.forEach { it.validate() } }
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is TaxRate &&
            this.description == other.description &&
            this.totalTaxRate == other.totalTaxRate &&
            this.effectiveTaxRate == other.effectiveTaxRate &&
            this.company == other.company &&
            this.remoteWasDeleted == other.remoteWasDeleted &&
            this.id == other.id &&
            this.remoteId == other.remoteId &&
            this.fieldMappings == other.fieldMappings &&
            this.modifiedAt == other.modifiedAt &&
            this.remoteData == other.remoteData &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    description,
                    totalTaxRate,
                    effectiveTaxRate,
                    company,
                    remoteWasDeleted,
                    id,
                    remoteId,
                    fieldMappings,
                    modifiedAt,
                    remoteData,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "TaxRate{description=$description, totalTaxRate=$totalTaxRate, effectiveTaxRate=$effectiveTaxRate, company=$company, remoteWasDeleted=$remoteWasDeleted, id=$id, remoteId=$remoteId, fieldMappings=$fieldMappings, modifiedAt=$modifiedAt, remoteData=$remoteData, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var description: JsonField<String> = JsonMissing.of()
        private var totalTaxRate: JsonField<Double> = JsonMissing.of()
        private var effectiveTaxRate: JsonField<Double> = JsonMissing.of()
        private var company: JsonField<String> = JsonMissing.of()
        private var remoteWasDeleted: JsonField<Boolean> = JsonMissing.of()
        private var id: JsonField<String> = JsonMissing.of()
        private var remoteId: JsonField<String> = JsonMissing.of()
        private var fieldMappings: JsonValue = JsonMissing.of()
        private var modifiedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteData: JsonField<List<RemoteData>> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(taxRate: TaxRate) = apply {
            this.description = taxRate.description
            this.totalTaxRate = taxRate.totalTaxRate
            this.effectiveTaxRate = taxRate.effectiveTaxRate
            this.company = taxRate.company
            this.remoteWasDeleted = taxRate.remoteWasDeleted
            this.id = taxRate.id
            this.remoteId = taxRate.remoteId
            this.fieldMappings = taxRate.fieldMappings
            this.modifiedAt = taxRate.modifiedAt
            this.remoteData = taxRate.remoteData
            additionalProperties(taxRate.additionalProperties)
        }

        /** The tax rate's description. */
        fun description(description: String) = description(JsonField.of(description))

        /** The tax rate's description. */
        @JsonProperty("description")
        @ExcludeMissing
        fun description(description: JsonField<String>) = apply { this.description = description }

        /** The tax rate's total tax rate. */
        fun totalTaxRate(totalTaxRate: Double) = totalTaxRate(JsonField.of(totalTaxRate))

        /** The tax rate's total tax rate. */
        @JsonProperty("total_tax_rate")
        @ExcludeMissing
        fun totalTaxRate(totalTaxRate: JsonField<Double>) = apply {
            this.totalTaxRate = totalTaxRate
        }

        /** The tax rate's effective tax rate. */
        fun effectiveTaxRate(effectiveTaxRate: Double) =
            effectiveTaxRate(JsonField.of(effectiveTaxRate))

        /** The tax rate's effective tax rate. */
        @JsonProperty("effective_tax_rate")
        @ExcludeMissing
        fun effectiveTaxRate(effectiveTaxRate: JsonField<Double>) = apply {
            this.effectiveTaxRate = effectiveTaxRate
        }

        /** The company the tax rate belongs to. */
        fun company(company: String) = company(JsonField.of(company))

        /** The company the tax rate belongs to. */
        @JsonProperty("company")
        @ExcludeMissing
        fun company(company: JsonField<String>) = apply { this.company = company }

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        fun remoteWasDeleted(remoteWasDeleted: Boolean) =
            remoteWasDeleted(JsonField.of(remoteWasDeleted))

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        @JsonProperty("remote_was_deleted")
        @ExcludeMissing
        fun remoteWasDeleted(remoteWasDeleted: JsonField<Boolean>) = apply {
            this.remoteWasDeleted = remoteWasDeleted
        }

        fun id(id: String) = id(JsonField.of(id))

        @JsonProperty("id") @ExcludeMissing fun id(id: JsonField<String>) = apply { this.id = id }

        /** The third-party API ID of the matching object. */
        fun remoteId(remoteId: String) = remoteId(JsonField.of(remoteId))

        /** The third-party API ID of the matching object. */
        @JsonProperty("remote_id")
        @ExcludeMissing
        fun remoteId(remoteId: JsonField<String>) = apply { this.remoteId = remoteId }

        @JsonProperty("field_mappings")
        @ExcludeMissing
        fun fieldMappings(fieldMappings: JsonValue) = apply { this.fieldMappings = fieldMappings }

        /** This is the datetime that this object was last updated by Merge */
        fun modifiedAt(modifiedAt: OffsetDateTime) = modifiedAt(JsonField.of(modifiedAt))

        /** This is the datetime that this object was last updated by Merge */
        @JsonProperty("modified_at")
        @ExcludeMissing
        fun modifiedAt(modifiedAt: JsonField<OffsetDateTime>) = apply {
            this.modifiedAt = modifiedAt
        }

        fun remoteData(remoteData: List<RemoteData>) = remoteData(JsonField.of(remoteData))

        @JsonProperty("remote_data")
        @ExcludeMissing
        fun remoteData(remoteData: JsonField<List<RemoteData>>) = apply {
            this.remoteData = remoteData
        }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): TaxRate =
            TaxRate(
                description,
                totalTaxRate,
                effectiveTaxRate,
                company,
                remoteWasDeleted,
                id,
                remoteId,
                fieldMappings,
                modifiedAt,
                remoteData.map { it.toUnmodifiable() },
                additionalProperties.toUnmodifiable(),
            )
    }
}
